GSoC/2016/JuliaQuantum-:-Framework-for-QuComput.md

```diff
@@ -0,0 +1,414 @@
+- visualizations:
+    + **quantum circuit visualization** visualize the quantum circuits desribed by user in this framework by [Compose.jl](https://github.com/dcjones/Compose.jl) or [Luxor.jl](https://github.com/cormullion/Luxor.jl)
+    + **graph states** graph states visualization will be based on [GraphPlot.jl](https://github.com/afternone/GraphPlot.jl)
+
+At the end of the project, JuliaQuantum will be equipped with a framework to simulate quantum computation. Including the features about:
+
+- quantum circuit's simulation and visualization
+- Adiabatic model simulators
+- one-way model's simulation and visualization
+- a GPU based APIs for QuArrays and matrix for Hamiltonians(if the QuBase consider to add this feature and time is enough)
+
+The project will make use of existing packages and aims to contribute back, for example:
+
+- `QuArray` will be used to store the states of Qubits
+- QuDynamics.jl will provide adiabatic computational model with more efficient Shordinger equation solvers with good accuracy. 
+
+
+Explanation of nouns:
+
+- **computational models** here means the algorithms that can provide universal computation and can be realized in experiments
+
+
+---
+## Details :
+
+### Structure of Types
+
+![](imag/AbstractGate.png)
+
+
+### Prototype
+I wrote a prototype for my idea with only simple features about quantum circuit and adiabatic quantum simulation module in [Kabab](https://github.com/Roger-luo/Kebab).
+
+### Implementation:
+
+#### Type System
+- **Abstract Gates** `AbstractGates` is a super type for all modules in the structure of this framework. The simulations of a quantum computation model will be construct in a quantum-circuit-like way. The other quantum computation model will be a gate-like module(subtype of `AbstractGates`) in quantum circuit.
+- **Quantum Gates** quantum gates includes gate-like modules and single qubit gates like Hadamard gate,Pauli gates...
+- **Graph** graph type will be used to store graph states for one-way model, the type is current considered to use adjacent matrix and use [GraphPlot.jl](https://github.com/afternone/GraphPlot.jl) as dependency to visualize
+- **Adiabatic System** `AdiaSystem` is the a basic type for adiabatic model's Hamiltonian(This may be reconstructed to the Hamiltonian type in QuDynamics, but I'm not sure which one is better)
+
+- *I still do not have a clear thought of quantum error corrections types in the type system. The current plan is making encoding modules become a quantum gate-like module*
+
+At the end of the project, JuliaQuantum will be equipped with a framework to simulate quantum computation models including quantum circuits, adiabatic quantum computation, one-way model, and quantum error corrections with CSS code, graph code.
+
+
+---
+### Design :
+
+The idea is to organize solvers,quantum gates and other quantum computation simulators in the form of quantum gates.
+
+```julia
+abstract AbstractGates
+abstract AbstractCircuit
+
```

Comment:
AbstractGates include computational models cause other computational models is considered to be a kind of quantum gate that input a state and output a state, which acts like a gate in the program's structure. So I called it AbstractGates, cause it will include all the gate-like functions/matrixes